### PR TITLE
[expo-keep-awake] Add web implementation for expo-keep-awake

### PR DESCRIPTION
**Pull request by jarvisluong**
# Why

As mentioned in https://github.com/expo/expo/issues/6889

# How

I implemented as the implementation guide of https://web.dev/wakelock/

Since the wake lock request in web will return a Promise, I changed the public API of expo-keep-awake to reflect that also.

# Test Plan

To be updated

### CODE REVIEW

**Comment by jarvisluong** [comment]
I will update the docs and change log as well if we are okay with the public API changes.

File: packages/expo-keep-awake/src/ExpoKeepAwake.ts [hunk 1]
```diff
@@ -1,2 +1,7 @@
 import { NativeModulesProxy } from '@unimodules/core';
+
+export function isAvailable() {
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/ExpoKeepAwake.ts:3]: This must be implemented in native code.

File: packages/expo-keep-awake/src/ExpoKeepAwake.web.ts [hunk 2]
```diff
@@ -1 +1,36 @@
-export default {};
+import { KeepAwakeEventType } from './ExpoKeepAwake.types';
+
+export function isAvailable() {
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/ExpoKeepAwake.web.ts:3]:
```suggestion
export function isAvailableAsync(): Promise<boolean> {
```

There should only be one `default export {}` with all "native" methods inside of it.

File: packages/expo-keep-awake/src/ExpoKeepAwake.web.ts [hunk 3]
```diff
@@ -1 +1,36 @@
...
+      if (wakeLockMap[tag]) {
+        wakeLockMap[tag].addEventListener('release', () => {
+          listener(KeepAwakeEventType.RELEASE);
+        });
+      }
+    },
+  };
+}
+
+export default createKeepAwakeManager();
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/ExpoKeepAwake.web.ts:36]: Just export a default object like all of the other modules.

File: packages/expo-keep-awake/src/index.ts [hunk 4]
```diff
@@ -1,20 +1,41 @@
...
+  }
+}
+
+export {
+  addEventListener,
+  activateKeepAwake,
+  deactivateKeepAwake,
+  useKeepAwake,
+  KeepAwakeEventType,
+};
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/index.ts:41]: Try to keep the export pattern similar to the other modules https://github.com/expo/expo/blob/master/packages/expo-battery/src/Battery.ts

All functions should be exported inline, all types can be exported at the end of the file.

File: packages/expo-keep-awake/src/ExpoKeepAwake.web.ts [hunk 5]
```diff
@@ -1 +1,36 @@
...
+
+      const wakeLock = await (navigator as any).wakeLock.request('screen');
+      wakeLockMap[tag] = wakeLock;
+    },
+    async deactivate(tag: string) {
+      if (wakeLockMap[tag]) {
+        wakeLockMap[tag].release();
+        wakeLockMap[tag] = null;
+      } else {
+        throw new Error(`The wake lock with tag ${tag} has not activated yet`);
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/ExpoKeepAwake.web.ts:23]:
```suggestion
        throw new CodedError('ERR_KEEP_AWAKE_TAG_INVALID', 'The wake lock with tag ${tag} has not activated yet');
```

File: packages/expo-keep-awake/src/index.ts [hunk 6]
```diff
@@ -1,20 +1,41 @@
...
+    };
   }, [tag]);
 }
 
-export function activateKeepAwake(tag: string = ExpoKeepAwakeTag): void {
-  if (ExpoKeepAwake.activate) ExpoKeepAwake.activate(tag);
+async function activateKeepAwake(tag: string = ExpoKeepAwakeTag): Promise<void> {
+  if (isAvailable() && ExpoKeepAwake.activate) {
+    await ExpoKeepAwake.activate(tag);
+  }
```
Comment (EvanBacon) [inline packages/expo-keep-awake/src/index.ts:20]:
```suggestion
  if (!(await isAvailableAsync())) {
    if (__DEV__) {
      console.warn('KeepAwake.activateKeepAwake() was invoked on a device that doesn't support the KeepAwake API.');
    }
    return;
  }
  if (ExpoKeepAwake.activate) {
    await ExpoKeepAwake.activate(tag);
  }
```

Let's warn in dev mode that the API has been misused (we do this in AuthSession API)

Comment (EvanBacon) [reply packages/expo-keep-awake/src/index.ts:20]: Apply the same to all methods

**Review by EvanBacon — CHANGES_REQUESTED** [review]
There are some core API changes that need to be changed in order to match the ecosystem.

**Comment by jarvisluong** [comment]
> There are some core API changes that need to be changed in order to match the ecosystem.

could you elaborate which core API should be changed. By "ecosystem" do you mean this particular package?

**Comment by EvanBacon** [comment]
ya, any of them but I linked a module in the comments that you can copy https://github.com/expo/expo/blob/master/packages/expo-battery/src/Battery.ts

**Comment by jarvisluong** [comment]
@EvanBacon Hi! Any updates on this PR? (Sorry for disturbing 😬)

**Comment by Nfinley** [comment]
This web keep awake feature would be incredible to have. Any updates on this? Anything I can do to help? @EvanBacon 

Thank you for all your work on expo

**Comment by brentvatne** [comment]
replaced by https://github.com/expo/expo/pull/15826

**Pull request closed without merging**